### PR TITLE
Enumerable methods

### PR DESCRIPTION
**Pull request by lucasmazo32**
The instructions to use the file are:

- Select the method you want to use (They work the same as the original ones, even with empty arrays)
  - `my_each, my_map, my_select, etc`
- You can use the same array as the one I have in the program or modify it.
- Run the program using ruby 
  - `ruby enumerable.rb`

I tried to replicate as close as I could the functionalities of the Enumerable methods, even with none? all? or any? in the best way possible and taking into consideration empty arrays.

I hope you enjoy the code!

Lucas Mazo

### CODE REVIEW

**Review by Redvanisation — CHANGES_REQUESTED** [review]
# Changes required  ♻️ 

### Hello team  🌟

**Good work on this project, but we need to fix the issues below before moving on ⬇️** 

- The methods `my_each`, `my_each_with_index`, `my_select` and `my_map` aren't working correctly, the case when there is no argument given to the method should be handled, consider using the `block_given?` method, check out this [page](https://ruby-doc.org/core-2.6.5/Enumerable.html) for reference.

### _After making the required changes please submit a new code review request_

### Good luck! 🍀

**Review by Redvanisation — CHANGES_REQUESTED** [review]
# Changes required  ♻️ 

### Hello @lucasmazo32   🌟

**Great work Implementing the changes so far, but we still have to fix the following issues ⬇️** 

**The below methods are behaving correctly when I check it against the original method I get the following results:**

1. `my_all?` 

> 2.6.3 :014 > [1, true, 'hi', []].my_all? == [1, true, 'hi', []].all?
>  => false 

> [1,2,3,4].my_all?(Integer)
> ArgumentError (wrong number of arguments (given 1, expected 0))

> 2.6.3 :012 > [1,2,3,4].all?(Integer)
>  => true 

2. `my_any?` 

> 2.6.3 :032 > [1, true, 'hi', []].my_any? == [1, true, 'hi', []].any?
>  => false 

> 2.6.3 :012 > [1,2,3,4].my_any?(Integer)
> ArgumentError (wrong number of arguments (given 1, expected 0))

3. `my_none?`

> 2.6.3 :012 > [1,2,3,4].my_none?(String)
> ArgumentError (wrong number of arguments (given 1, expected 0))

4. `my_inject` isn't working as intended it should also give the same result as the original `inject` method.

Please refer to this [page](https://ruby-doc.org/core-2.6.5/Enumerable.html) for info about the original methods then fix the issues.

### _After making the required changes please submit a new code review request_

### Good luck! 🍀

**Comment by shshamim63** [comment]
## Status :recycle:  [!! Required Changes !!] :recycle: 
___
Hi @lucasmazo32 :wave: ,
So far awesome job :clap: . You have covered most of the part asked by the previous reviewer. But some are not working as the actual one. There are some additional cases that the previous reviewer forgot to inform(:disappointed: Really sorry for that.). Believe me those check are really necessary to make you enumerable methods behave properly as the original one. I am sure that you will cover them all.
___
#### Please follow the instructions below :point_down:  to complete this project.
___
### :space_invader: [Mandatory Changes]
___
1. [**my_all?**](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-all-3F)
...1. When no block or argument is given returns true when none of the collection members are false or nil. You have covered the case for nil. What about the false?
`[1, 2, 'Hello', false].my_all?` should return false.
...2. When a Regex is passed as an argument returns true if all of the collection matches the Regex
`%w[welcome to the ruby world].my_all?(/d/)` should return true
...3. When a pattern other than Regex or a Class is given returns true if all of the collection matches the pattern.
`[3, 3, 3, 3].my_all?(3)` should return true, and `[1, 2, 3, 3].my_all?(3)` should return false.
___
2. [**my_any?**](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-any-3F)
...1. When no block or argument is given returns true if at least one of the collection is not false or nil 
`[1, nil, false].my_any?` should return true and `[nil, false].my_any?` should return false..
..2. When a Regex is passed as an argument returns false if none of the collection matches the Regex
`%w[welcome to the ruby world].my_any?(/d/)` should return true
...3. When a pattern other than Regex or a Class is given returns false if none of the collection matches the pattern
`[1, 2, 3, 3].my_any?(3)` should return true.
___
3. [**my_none?**](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-none-3F)
...1. When a Regex is passed as an argument returns true only if none of the collection matches the Regex
`%w[welcome to the ruby world].my_none?(/d/)` should return false
...2. When a pattern other than Regex or a Class is given returns true only if none of the collection matches the pattern 
`[1, 2, 3, 3].my_none?(3)` should return false.
4. [**my_inject**](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-inject)
The original method actually takes two parameter(initial, symbol). If you do not explicitly specify an initial value for memo, then the first element of collection is used as the initial value of memo. If you specify the symbol then the block is not 
...1. Combines all elements of enum by applying a binary operation, specified by a block
`(5..10).my_inject { |sum, n| sum + n }` should return 45
`(5..10).my_inject(5) { |sum, n| sum + n }` should return 50
...2. When a symbol is specified combines each element of the collection by applying the symbol as a named method
`(5..10).my_inject(5, :+)` should return 50
___
### :cake: [Optional Changes]
___
The name of the file ruby.rb is indicating that it contains some ruby code, which can be derived from the extension of the file. But to be more specific I think you can name it as enumerables.rb. Will give it a little push to guess what kind of task could be covered by this file. :smile: 
___
#### Please fix the above :point_up:  mentioned issues and ask for another review.
___
## :herb: Happy Coding :herb: 
### Shakhawat Hossain
[Linkedin](https://www.linkedin.com/in/shakhawathossainshamim/) | [Slack](__Shakhawat Hossain__)

**Review by Redvanisation — CHANGES_REQUESTED** [review]
# Changes required  ♻️ 

**Great work so far, but the following still needs to be fixed:**

1. `my_all?` This method is still not behaving like the original one:

> 2.6.3 :027 > [1, false, 'hi', []].my_all? == [1, false, 'hi', []].all?
>  => false 

- When a Regex is passed as an argument returns `true` if all of the collection matches the Regex but it's throwing the following errors instead:

> %w[dog door rod blade].my_all?(/d/) == %w[dog door rod blade].all?(/d/)
>      TypeError:
>        class or module required

and the same exact error when you try this: 

> [1,2,3,4].my_all?(3) == [1,2,3,4].all?(3)
>      TypeError:
>        class or module required

2. `my_any?` and `my_none?` The same thing for those methods also they aren't equal to the original methods and they throw a `TypeError: class or module required`.

3. `my_inject` when I call it with the following I get an error also:

> [0,3,5,1,9].my_inject(:+) == [0,3,5,1,9].inject(:+)
> "Error!"

### _After making the required changes please submit a new code review request_

### Good luck! 🍀

File: enumerable.rb [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+        cons = if i.zero?
+                 x
+               else
+                 cons.method(opposym).call(x)
+               end
+      end
+      return cons
+    end
+    opposym
+  end
```
Comment (whiz25) [inline enumerable.rb]: - [ ] identical and returns the same thing as ruby's [`inject`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-inject). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_inject.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
- [ ] Your method has an error that needs fixing;
inject

**Review by whiz25 — CHANGES_REQUESTED** [review]
## Amazing job :+1: 
You have implemented most of the requirements for the project. However, we can not merge this yet. You have one failing method. Please check my comment and fix the error.

### Martin Afani

**Comment by lucasmazo32** [comment]
I checked the "My inject" method and for me, it works well. Maybe the code reviewer was trying to use the inject method for an Integer, not an array, thus showing that problem. Since no further information was given, I will just upload a screenshot proving that my method works.

image

**Comment by shshamim63** [comment]
## Status :rabbit: [!! Required Changes !!] :rabbit: 
___
Hi @lucasmazo32 :wave: ,
Impressive work :clap: . All the methods are working exactly the original one except inject. Well there is an exceptional case that inject can perform. If you just clone the case , your enumerable will cover every edge cases. I believe that you will cover it easily.
___
### :spades: Mandatory Changes :spades: 
___
[**my_inject**](https://ruby-doc.org/core-2.6/Enumerable.html#method-i-inject)
Suppose, an array of string is given
`word = %w[dog cat ball blade]`
A proc is also generated 
`search = Proc.new { |memo, word| memo.length > word.length ? memo : word }`
After executions of `word.my_inject(&search)` Should return 'blade'. 
:hibiscus: Hint : Well I think is better to check, what parameter is containing. :thinking:  
___
#### Please fix the above mentioned issue and send a review request. 
___
## :herb: Happy Coding :herb: 
### Shakhawat Hossain
[Linkedin](https://www.linkedin.com/in/shakhawathossainshamim/) | [Slack](__Shakhawat Hossain__)

**Review by Redvanisation — APPROVED** [review]
# Project approved ✔️ 

### Hello @lucasmazo32  🌟 

### _Nice work on this project, it seems that you have met the requirements so I am approving your project. 👍_ 

### You may merge to master and submit a project completion form. 🚀 

### Good luck! 🍀

**Pull request merged**